app/services/profanity_detector.rb

```diff
@@ -0,0 +1,56 @@
+# Detects profanity in a string, the list of profane words
+# comes from the profanity_filter gem for now
+module ProfanityDetector
+  module_function
+
+  # Calls a block until the result does not contain profanity
+  # @yield block that generates a random value
+  # @yieldreturn [String] string to check for profanity
+  def without_profanity
+    loop do
+      word = yield
+
+      return word unless profane?(word)
+    end
+  end
+
+  # Returns true if the string contains profanity inside it
+  # ProfanityFilter::Base.profane? splits by word, but this
+  # checks all substrings
+  def profane?(str)
+    preprocess_if_needed!
+
+    str_chars = str.gsub(/\W/, '').downcase.chars
+
+    (min_profanity_length..[str.length, max_profanity_length].min).each do |size|
+      profane_words = @profanity_by_length[size]
+      next if profane_words.empty?
+
+      str_chars.each_cons(size) do |letters|
```

Comment:
Is it necessary to chunk up the string characters like this, or could we just do an `include?` on the normalized version of the string?

```rb
def profane?(str)
  preprocess_if_needed!

  normalized_str = str.gsub(/\W/, '').downcase

  (min_profanity_length..[normalized_str.length, max_profanity_length].min).any? do |size|
    profane_words = @profanity_by_length[size]
    profane_words.any? { |profane_word| normalized_str.include?(profane_word) }
  end
end
```

Reply:
I don't feel strongly either way, but it's slightly faster to do it as the current implementation has it

Reply:
> I don't feel strongly either way, but it's slightly faster to do it as the current implementation has it

Is it? I'd ran some light benchmarks earlier and it appeared to be faster. I adapted them to use the `benchmark-ips` library you've been using and it still appears faster.

https://gist.github.com/aduth/8d1b0e0a251d21c63c29665c706df636

```
Warming up --------------------------------------
              before    18.116k i/100ms
               after    28.235k i/100ms
Calculating -------------------------------------
              before    176.544k (± 3.1%) i/s -    887.684k in   5.033314s
               after    277.081k (± 7.8%) i/s -      1.384M in   5.050753s
```

🤷

Reply:
I lost track of which parts we benchmarked against what, but I agree with @mitchellhenke that the current code is faster than the proposal to do `Set#any?`. This is a good exercise in keeping me honest so let me write out my thinking:

- `Set#include?(String)` is roughly O(1) since it's a direct membership check
- `String#include?(String)` is an O(N) check since internally it has to 
- `Set#any? { String#include?(String))` is gonna be O(N<sup>2</sup>) because its iterating over each set element to do `String#include?(String)`

### Current code
`[min...max].each { |size| str.chars.each_cons(size) }`
for a 5-character input, that is 3 outer loop iterations, (3, 4, 5)
- 3 substrings at 3 chars (abc, bcd, cde)
- 2 substrings at 4 chars (abcd, bcde)
- 1 substrings at 5 chars (abcde)

for a total of 6 checks, each check an O(1) check of set inclusion

### Proposed code
`[min..max].each { |size| sets[size].any? { |word| str.include?(word) }`
for a 5-character input, that is 3 outer loop iterations (3, 4, 5)
- 7 words at 3 chars
- 26 words at 4 chars
- 27 words at 5 chars

for a total of 59 checks, each check is an O(N) String#include?(String), even if that was a faster check there are still more of them


---

For reference, here's breakdown of how many profane words @ each length
```ruby
ProfanityDetector.instance_variable_get(:@profanity_by_length).transform_values(&:size).sort_by(&:first).to_h
=> {3=>7,
 4=>26,
 5=>26,
 6=>27,
 7=>30,
 8=>22,
 9=>7,
 10=>15,
 11=>20,
 12=>11,
 13=>5,
 14=>1}
```

Reply:
In the interest of bike shedding, I re-ran it, and it the difference is negligible when generating random codes, and Andrew's proposed version is slightly faster when excluding code generation. Given it's simpler and potentially faster, I'd slightly favor Andrew's suggestion.

```ruby
def profane_old?(str)
  preprocess_if_needed!

  str_chars = str.gsub(/\W/, '').downcase.chars

  (min_profanity_length..[str_chars.length, max_profanity_length].min).each do |size|
    profane_words = @profanity_by_length[size]
    next if profane_words.empty?

    str_chars.each_cons(size) do |letters|
     return true if profane_words.include?(letters.join)
    end
  end

  false
end

def profane_new?(str)
  preprocess_if_needed!

  normalized_str = str.gsub(/\W/, '').downcase

  (min_profanity_length..[normalized_str.length, max_profanity_length].min).any? do |size|
    profane_words = @profanity_by_length[size]
    profane_words.any? { |profane_word| normalized_str.include?(profane_word) }
  end
end

random_codes = (1..10_000).map { |x| Base32::Crockford.encode(SecureRandom.random_number(2**(5 * 10)), length: 10) }

Benchmark.ips do |x|
  x.time = 120
  x.report("proposed") { random_codes.each { |x| ProfanityDetector.profane_new?(x) } }
  x.report("current") { random_codes.each { |x| ProfanityDetector.profane_old?(x) } }
  x.compare!
end

Warming up --------------------------------------
            proposed     1.000  i/100ms
             current     1.000  i/100ms
Calculating -------------------------------------
            proposed      3.640  (± 0.0%) i/s -    437.000  in 120.220159s
             current      3.262  (± 0.0%) i/s -    391.000  in 120.164345s

Comparison:
            proposed:        3.6 i/s
             current:        3.3 i/s - 1.12x  (± 0.00) slower
```